From: Aaron <aaronsb@gmail.com>
Date: Mon, 25 Nov 2024 10:00:00 +0000
Subject: [PATCH] net: stmmac: dwmac4: use blocking DMA reset for RK3568

Use blocking mdelay() instead of readl_poll_timeout() for DMA reset
to ensure the MAXIO PHY has time to stabilize before GMAC tries to
use it. This prevents "Failed to reset the dma" errors on RK3568.

Signed-off-by: Aaron <aaronsb@gmail.com>
---
 drivers/net/ethernet/stmicro/stmmac/dwmac4_lib.c | 12 ++++++++----
 1 file changed, 8 insertions(+), 4 deletions(-)

diff --git a/drivers/net/ethernet/stmicro/stmmac/dwmac4_lib.c b/drivers/net/ethernet/stmicro/stmmac/dwmac4_lib.c
index 0d185e54e..ce8e9eff4 100644
--- a/drivers/net/ethernet/stmicro/stmmac/dwmac4_lib.c
+++ b/drivers/net/ethernet/stmicro/stmmac/dwmac4_lib.c
@@ -16,14 +16,18 @@
 int dwmac4_dma_reset(void __iomem *ioaddr)
 {
 	u32 value = readl(ioaddr + DMA_BUS_MODE);
+	int limit = 20;

 	/* DMA SW reset */
 	value |= DMA_BUS_MODE_SFT_RESET;
 	writel(value, ioaddr + DMA_BUS_MODE);

-	return readl_poll_timeout(ioaddr + DMA_BUS_MODE, value,
-				 !(value & DMA_BUS_MODE_SFT_RESET),
-				 10000, 1000000);
+	while (limit--) {
+		if (!(readl(ioaddr + DMA_BUS_MODE) & DMA_BUS_MODE_SFT_RESET))
+			return 0;
+		mdelay(10);
+	}
+	return -EBUSY;
 }

 void dwmac4_set_rx_tail_ptr(struct stmmac_priv *priv, void __iomem *ioaddr,
